_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+states from each initial program state. Methods more closely match the
+operations performed by actual computing hardware, and are capable of
+describing non-deterministic or non-terminating computations. They are
+also generally meant to be compiled to executable programs. However,
+reasoning about methods is somewhat different than reasoning about
+functions. You can prove that a method agrees with a single contract,
+written as `requires`, `ensures`, `modifies`, and `decreases` clauses,
+but you generally do not expand the body of a method as part of
+performing a separate proof (though it is sometimes possible to do so,
+as will be described in a future post).
+
+These differing attributes come into play particularly starkly when
+developing programs that have either or both of the following
+attributes:
+
+* The most natural specification of an API involves describing the
+_relationship_ between the available operations, rather than
+individual descriptions of each operation in isolation.
+
+* There's more than one way to implement a particular bit of
+functionality and, due to the tradeoffs between the available
+approaches, you don't want to choose just one. Perhaps you'll use
+one in some circumstances and another in other circumstances.
+
+In this situation, it is particularly valuable to clearly
+distinguish between the specification and implementation of a
+program, and to restrict proofs about the implementation to those
+that guarantee agreement with the specification, performing all
+other reasoning on the specification. The rest of this post will
+explore a simple library that has both of the attributes listed
+above: a specification and implementation of the well-known stack
+data type.
+
+The (arguably) most concise and natural specification of a stack
+comes from a description of a relationship between the push and pop
+operations. Namely, if you push an item onto a stack, and then pop
+an item off of the resulting stack, you should get back the item you
+originally pushed. Similarly, the value of the stack itself after
+these two operations should be unchanged: when you add something to
+the top and immediately remove it, the rest of the stack stays the
+same.
+
+On the implementation side, there are multiple ways to build a
+stack. Common implementation techniques include custom linked lists,
+consecutive arrays (reallocated when full), and the re-use of
+built-in data types with appropriate properties, such as sequences
+or vectors. Perhaps you wouldn't use more than one of these in a
+given application, but we can still look at how to straightforwardly
+show that multiple implementations, using different techniques, all
+fit the same notion of what it means to be a stack.
+
+To start with, let's specify what it means to be a stack in the
+following module.
+
+{% highlight dafny %}
+module StackSpecification {
+  // Most clients, including clients of implementations, get no details.
```

Comment:
Please try to put this comments after the Dafny code.

Reply:
I'm not sure I agree. I find it more readable before the code, and I think it's a more common approach, and therefore less surprising to readers.